src/frontend/src/flows/authenticate/validateDerivationOrigin.ts

```diff
@@ -0,0 +1,76 @@
+import { Principal } from "@dfinity/principal";
+
+export type ValidationResult =
+  | { result: "valid" }
+  | { result: "invalid"; message: string };
+
+export const validateDerivationOrigin = async (
+  authRequestOrigin: string,
+  derivationOrigin?: string
+): Promise<ValidationResult> => {
+  if (
+    derivationOrigin === undefined ||
+    derivationOrigin === authRequestOrigin
+  ) {
+    // this is the default behaviour -> no further validation necessary
+    return { result: "valid" };
+  }
+
+  // check format of derivationOrigin
+  const matches = /^https:\/\/([\w-]*)(\.raw)?\.ic0\.app$/.exec(
+    derivationOrigin
+  );
+  if (matches === null) {
+    return {
+      result: "invalid",
+      message:
+        'derivationOrigin does not match regex "^https:\\/\\/([\\w-]*)(\\.raw)?\\.ic0\\.app$"',
+    };
+  }
+
+  try {
+    const canisterId = Principal.fromText(matches[1]); // verifies that a valid canister id was matched
+    const alternativeOriginsUrl = `https://${canisterId.toText()}.ic0.app/.well-known/ii-alternative-origins`;
+    const response = await fetch(
+      // always fetch non-raw
+      alternativeOriginsUrl,
+      // fail on redirects
+      { redirect: "error" }
+    );
+
+    if (response.status !== 200) {
+      return {
+        result: "invalid",
+        message: `resource ${alternativeOriginsUrl} returned invalid status: ${response.status}`,
+      };
+    }
+
+    const alternativeOriginsObj = (await response.json()) as {
+      alternativeOrigins: string[];
+    };
+
+    // check for expected property
+    if (!Array.isArray(alternativeOriginsObj?.alternativeOrigins)) {
+      return {
+        result: "invalid",
+        message: `resource ${alternativeOriginsUrl} has invalid format: received ${alternativeOriginsObj}`,
+      };
+    }
+
+    // check allowed alternative origins
+    if (!alternativeOriginsObj.alternativeOrigins.includes(authRequestOrigin)) {
```

Comment:
looks like we're trying to handle malformed content. Should also display a nice error if the content is not JSON?

Reply:
I think the current error message is clear enough.

Example for the content `not-json`:
```
JSON.parse: unexpected keyword at line 1 column 1 of the JSON data
```